presto-hive/src/test/java/com/facebook/presto/hive/parquet/BenchmarkParquetPageSource.java

```diff
@@ -0,0 +1,528 @@
+                }
+
+                if (!conjunction.isPresent()) {
+                    conjunction = rowExpression;
+                }
+                else {
+                    conjunction = Optional.of(specialForm(
+                            AND,
+                            BOOLEAN,
+                            conjunction.get(),
+                            rowExpression.get()));
+                }
+            }
+            return conjunction;
+        }
+
+        public List<Optional<Page>> process(Page page)
+        {
+            return ImmutableList.copyOf(
+                    pageProcessor.process(
+                            null,
+                            new DriverYieldSignal(),
+                            newSimpleAggregatedMemoryContext().newLocalMemoryContext(PageProcessor.class.getSimpleName()),
+                            page));
+        }
+
+        private Type getTypeFromTypeSignature()
+        {
+            switch (typeSignature) {
+                case "boolean":
+                    return BOOLEAN;
+                case "int32":
+                    return INTEGER;
+                case "int64":
+                    return BIGINT;
+                case "int96":
+                    return TIMESTAMP;
+                case "varchar":
+                    return VarcharType.createVarcharType(400);
+                default:
+                    throw new UnsupportedOperationException("Unsupported type: " + typeSignature);
+            }
+        }
+
+        private List<RowExpression> getProjections(Type type)
+        {
+            ImmutableList.Builder<RowExpression> builder = ImmutableList.builder();
+            for (int i = 0; i < channelCount; i++) {
+                builder.add(new InputReferenceExpression(i, type));
+            }
+
+            return builder.build();
+        }
+
+        private Optional<RowExpression> getFilter(int columnIndex, Type type, float filterRate, boolean filterAllowNull, float selectionRateForNonNull)
+        {
+            if (filterRate == NO_FILTER) {
+                return Optional.empty();
+            }
+
+            RowExpression filter = null;
+
+            if (BOOLEAN.equals(type)) {
+                filter = call(EQUAL.name(),
+                        functionManager.resolveOperator(EQUAL, fromTypes(type, type)),
+                        BOOLEAN,
+                        field(columnIndex, type),
+                        constant(true, type));
+            }
+
+            if (INTEGER.equals(type)) {
+                filter = specialForm(AND,
+                        BOOLEAN,
+                        call(GREATER_THAN_OR_EQUAL.name(),
+                                functionManager.resolveOperator(GREATER_THAN_OR_EQUAL, fromTypes(type, type)),
+                                BOOLEAN,
+                                field(columnIndex, type),
+                                constant((long) (Integer.MIN_VALUE * selectionRateForNonNull), type)),
```

Comment:
@ClarenceThreepwood It looks good, will you be able to post some sample output row counts for the final output pages for the following cases?
 1. no nulls boolean
 2. partial nulls boolean
 3. no nulls int
 4. partial nulls int
Thank you!

Reply:
Total Row Count = 10^6


					      0.1|-1				        0.1|0.5

boolean (no nulls)			8182304				2727065

boolean (partial nulls)		8999746				4500412

int32 (no nulls) 			9000758				4499316

int32 (partial nulls)			9000191				4500421

Reply:
Thanks @ClarenceThreepwood ! The boolean (no nulls) number for 0.1|0.5 doesn't seem right. Can you take a look? What is the output if there's just one column with filter rates 0.1, 0.5 and 0.9?

Reply:
I believe this is the intention and it is the same as in BenchmarkSelectiveReaders. 
For 0.1|0.5, the filter selectivity for booleans (no nulls) works out to 
c1 = (1-0.1)/(1+0.1) = 0.81
c2 = (1-0.5)/(1+0.5) = 0.33
Final count of output rows = c1*c2 = **0.267**


boolean (no nulls)
filterRate | outputRows
--- | ---
0.1 | 8182304
0.5 | 3334245
0.9 | 526971

Reply:
Updated the boolean generation as discussed. The new numbers are 

Total Row Count = 10^6


|filterRate | 0.1\|-1|  0.1\|0.5|
| --- | --- | --- |
|boolean        (no nulls)                |                  8999659      |                   4497122|
|boolean (partial nulls)     |       8999746       |                4500412|
|int32           (no nulls)           |            9000758	   |                    4499316|
|int32     (partial nulls)	      |         9000191	          |            4500421|


boolean (no nulls)
filterRate | outputRows
--- | ---
0.1 | 8999659
0.5 | 5000426
0.9 | 1000889